File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/render/RenderSessionHandle.java
package org.maplibre.nativeffi.render;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Function;
import org.maplibre.nativeffi.error.InvalidArgumentException;
import org.maplibre.nativeffi.error.MaplibreStatus;
import org.maplibre.nativeffi.geo.Feature;
import org.maplibre.nativeffi.internal.access.InternalAccess;
import org.maplibre.nativeffi.internal.c.MapLibreNativeC;
import org.maplibre.nativeffi.internal.c.mln_metal_owned_texture_frame;
import org.maplibre.nativeffi.internal.c.mln_vulkan_owned_texture_frame;
import org.maplibre.nativeffi.internal.lifecycle.HandleState;
import org.maplibre.nativeffi.internal.loader.NativeAccess;
import org.maplibre.nativeffi.internal.memory.MemoryUtil;
import org.maplibre.nativeffi.internal.status.Status;
import org.maplibre.nativeffi.internal.struct.CoreStructs;
import org.maplibre.nativeffi.internal.struct.QueryStructs;
import org.maplibre.nativeffi.internal.struct.RenderStructs;
import org.maplibre.nativeffi.internal.struct.ValueStructs;
import org.maplibre.nativeffi.json.JsonValue;
import org.maplibre.nativeffi.map.MapHandle;
import org.maplibre.nativeffi.query.FeatureExtensionResult;
import org.maplibre.nativeffi.query.FeatureStateSelector;
import org.maplibre.nativeffi.query.QueriedFeature;
import org.maplibre.nativeffi.query.RenderedFeatureQueryOptions;
import org.maplibre.nativeffi.query.RenderedQueryGeometry;
import org.maplibre.nativeffi.query.SourceFeatureQueryOptions;

/** Owned native render session handle. Close it on the map owner thread. */
public final class RenderSessionHandle implements AutoCloseable {
  private final MapHandle map;
  private final HandleState state;

  private RenderSessionHandle(MapHandle map, MemorySegment handle) {
    this.map = Objects.requireNonNull(map, "map");
    this.state = new HandleState("RenderSessionHandle", handle, map);
  }

  public static RenderSessionHandle attachOwnedTexture(
      MapHandle map, OwnedTextureDescriptor descriptor) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(map, "map");
    Objects.requireNonNull(descriptor, "descriptor");
    try (var arena = Arena.ofConfined()) {
      var outSession = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_owned_texture_attach(
              map.nativeHandle(InternalAccess.INSTANCE),
              RenderStructs.ownedTextureDescriptor(descriptor, arena),
              outSession));
      return new RenderSessionHandle(map, outSession.get(ValueLayout.ADDRESS, 0));
    }
  }

  public static RenderSessionHandle attachMetalOwnedTexture(
      MapHandle map, MetalOwnedTextureDescriptor descriptor) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(map, "map");
    Objects.requireNonNull(descriptor, "descriptor");
    try (var arena = Arena.ofConfined()) {
      var outSession = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_metal_owned_texture_attach(
              map.nativeHandle(InternalAccess.INSTANCE),
              RenderStructs.metalOwnedTextureDescriptor(descriptor, arena),
              outSession));
      return new RenderSessionHandle(map, outSession.get(ValueLayout.ADDRESS, 0));
    }
  }

  public static RenderSessionHandle attachMetalBorrowedTexture(
      MapHandle map, MetalBorrowedTextureDescriptor descriptor) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(map, "map");
    Objects.requireNonNull(descriptor, "descriptor");
    try (var arena = Arena.ofConfined()) {
      var outSession = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_metal_borrowed_texture_attach(
              map.nativeHandle(InternalAccess.INSTANCE),
              RenderStructs.metalBorrowedTextureDescriptor(descriptor, arena),
              outSession));
      return new RenderSessionHandle(map, outSession.get(ValueLayout.ADDRESS, 0));
    }
  }

  public static RenderSessionHandle attachVulkanOwnedTexture(
      MapHandle map, VulkanOwnedTextureDescriptor descriptor) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(map, "map");
    Objects.requireNonNull(descriptor, "descriptor");
    try (var arena = Arena.ofConfined()) {
      var outSession = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_vulkan_owned_texture_attach(
              map.nativeHandle(InternalAccess.INSTANCE),
              RenderStructs.vulkanOwnedTextureDescriptor(descriptor, arena),
              outSession));
      return new RenderSessionHandle(map, outSession.get(ValueLayout.ADDRESS, 0));
    }
  }

  public static RenderSessionHandle attachVulkanBorrowedTexture(
      MapHandle map, VulkanBorrowedTextureDescriptor descriptor) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(map, "map");
    Objects.requireNonNull(descriptor, "descriptor");
    try (var arena = Arena.ofConfined()) {
      var outSession = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_vulkan_borrowed_texture_attach(
              map.nativeHandle(InternalAccess.INSTANCE),
              RenderStructs.vulkanBorrowedTextureDescriptor(descriptor, arena),
              outSession));
      return new RenderSessionHandle(map, outSession.get(ValueLayout.ADDRESS, 0));
    }
  }

  public static RenderSessionHandle attachMetalSurface(
      MapHandle map, MetalSurfaceDescriptor descriptor) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(map, "map");
    Objects.requireNonNull(descriptor, "descriptor");
    try (var arena = Arena.ofConfined()) {
      var outSession = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_metal_surface_attach(
              map.nativeHandle(InternalAccess.INSTANCE),
              RenderStructs.metalSurfaceDescriptor(descriptor, arena),
              outSession));
      return new RenderSessionHandle(map, outSession.get(ValueLayout.ADDRESS, 0));
    }
  }

  public static RenderSessionHandle attachVulkanSurface(
      MapHandle map, VulkanSurfaceDescriptor descriptor) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(map, "map");
    Objects.requireNonNull(descriptor, "descriptor");
    try (var arena = Arena.ofConfined()) {
      var outSession = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_vulkan_surface_attach(
              map.nativeHandle(InternalAccess.INSTANCE),
              RenderStructs.vulkanSurfaceDescriptor(descriptor, arena),
              outSession));
      return new RenderSessionHandle(map, outSession.get(ValueLayout.ADDRESS, 0));
    }
  }

  public void resize(int width, int height, double scaleFactor) {
    NativeAccess.ensureLoaded();
    Status.check(
        MapLibreNativeC.mln_render_session_resize(state.requireLive(), width, height, scaleFactor));
  }

  public void renderUpdate() {
    NativeAccess.ensureLoaded();
    Status.check(MapLibreNativeC.mln_render_session_render_update(state.requireLive()));
  }

  public void detach() {
    NativeAccess.ensureLoaded();
    Status.check(MapLibreNativeC.mln_render_session_detach(state.requireLive()));
  }

  public void reduceMemoryUse() {
    NativeAccess.ensureLoaded();
    Status.check(MapLibreNativeC.mln_render_session_reduce_memory_use(state.requireLive()));
  }

  public void clearData() {
    NativeAccess.ensureLoaded();
    Status.check(MapLibreNativeC.mln_render_session_clear_data(state.requireLive()));
  }

  public void dumpDebugLogs() {
    NativeAccess.ensureLoaded();
    Status.check(MapLibreNativeC.mln_render_session_dump_debug_logs(state.requireLive()));
  }

  public void setFeatureState(FeatureStateSelector selector, JsonValue value) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(selector, "selector");
    Objects.requireNonNull(value, "value");
    try (var arena = Arena.ofConfined()) {
      Status.check(
          MapLibreNativeC.mln_render_session_set_feature_state(
              state.requireLive(),
              QueryStructs.featureStateSelector(selector, arena),
              ValueStructs.jsonValue(value, arena)));
    }
  }

  public JsonValue getFeatureState(FeatureStateSelector selector) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(selector, "selector");
    try (var arena = Arena.ofConfined()) {
      var outState = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_render_session_get_feature_state(
              state.requireLive(), QueryStructs.featureStateSelector(selector, arena), outState));
      return ValueStructs.jsonSnapshot(outState.get(ValueLayout.ADDRESS, 0))
          .orElseGet(() -> JsonValue.object(List.of()));
    }
  }

  public void removeFeatureState(FeatureStateSelector selector) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(selector, "selector");
    try (var arena = Arena.ofConfined()) {
      Status.check(
          MapLibreNativeC.mln_render_session_remove_feature_state(
              state.requireLive(), QueryStructs.featureStateSelector(selector, arena)));
    }
  }

  public List<QueriedFeature> queryRenderedFeatures(RenderedQueryGeometry geometry) {
    return queryRenderedFeaturesInternal(geometry, null, false);
  }

  public List<QueriedFeature> queryRenderedFeatures(
      RenderedQueryGeometry geometry, RenderedFeatureQueryOptions options) {
    return queryRenderedFeaturesInternal(
        geometry, Objects.requireNonNull(options, "options"), true);
  }

  public List<QueriedFeature> querySourceFeatures(String sourceId) {
    return querySourceFeaturesInternal(sourceId, null, false);
  }

  public List<QueriedFeature> querySourceFeatures(
      String sourceId, SourceFeatureQueryOptions options) {
    return querySourceFeaturesInternal(sourceId, Objects.requireNonNull(options, "options"), true);
  }

  public FeatureExtensionResult queryFeatureExtension(
      String sourceId, Feature feature, String extension, String extensionField) {
    return queryFeatureExtension(sourceId, feature, extension, extensionField, null);
  }

  public FeatureExtensionResult queryFeatureExtension(
      String sourceId,
      Feature feature,
      String extension,
      String extensionField,
      JsonValue arguments) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(feature, "feature");
    try (var arena = Arena.ofConfined()) {
      var outResult = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_render_session_query_feature_extensions(
              state.requireLive(),
              CoreStructs.stringView(Objects.requireNonNull(sourceId, "sourceId"), arena),
              ValueStructs.feature(feature, arena),
              CoreStructs.stringView(Objects.requireNonNull(extension, "extension"), arena),
              CoreStructs.stringView(
                  Objects.requireNonNull(extensionField, "extensionField"), arena),
              arguments == null ? MemorySegment.NULL : ValueStructs.jsonValue(arguments, arena),
              outResult));
      return QueryStructs.featureExtensionResult(outResult.get(ValueLayout.ADDRESS, 0));
    }
  }

  public TextureImageInfo textureImageInfo() {
    NativeAccess.ensureLoaded();
    try (var arena = Arena.ofConfined()) {
      var outInfo = MapLibreNativeC.mln_texture_image_info_default(arena);
      var status =
          MapLibreNativeC.mln_texture_read_premultiplied_rgba8(
              state.requireLive(), MemorySegment.NULL, 0, outInfo);
      var info = RenderStructs.textureImageInfo(outInfo);
      if (status == MaplibreStatus.OK.nativeCode()
          || (status == MaplibreStatus.INVALID_ARGUMENT.nativeCode() && info.byteLength() > 0)) {
        return info;
      }
      Status.check(status);
      throw new AssertionError("unreachable");
    }
  }

  /**
   * Reads into a caller-owned buffer.
   *
   * <p>Use {@link #textureImageInfo()} first when sizing reusable buffers; undersized buffers throw
   * {@link InvalidArgumentException} after native reports the required layout.
   */
  public TextureImageInfo readPremultipliedRgba8(NativeBuffer buffer) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(buffer, "buffer");
    synchronized (buffer) {
      try (var arena = Arena.ofConfined()) {
        var capacity = buffer.byteLength();
        var outInfo = MapLibreNativeC.mln_texture_image_info_default(arena);
        Status.check(
            MapLibreNativeC.mln_texture_read_premultiplied_rgba8(
                state.requireLive(),
                capacity == 0 ? MemorySegment.NULL : buffer.segment(),
                capacity,
                outInfo));
        return RenderStructs.textureImageInfo(outInfo);
      }
    }
  }

  public PremultipliedRgba8Image readPremultipliedRgba8() {
    var info = textureImageInfo();
    try (var buffer = NativeBuffer.allocate(info.byteLength())) {
      var readInfo = readPremultipliedRgba8(buffer);
      return new PremultipliedRgba8Image(
          readInfo.width(), readInfo.height(), readInfo.stride(), buffer.toByteArray());
    }
  }

  public void withMetalOwnedTextureFrame(Consumer<MetalOwnedTextureFrame> callback) {
    Objects.requireNonNull(callback, "callback");
    withMetalOwnedTextureFrame(
        frame -> {
          callback.accept(frame);
          return null;
        });
  }

  public <T> T withMetalOwnedTextureFrame(Function<MetalOwnedTextureFrame, T> callback) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(callback, "callback");
    try (var arena = Arena.ofConfined()) {
      var frameSegment = RenderStructs.metalOwnedTextureFrame(arena);
      Status.check(
          MapLibreNativeC.mln_metal_owned_texture_acquire_frame(state.requireLive(), frameSegment));
      var scope = new FrameScope();
      Throwable callbackFailure = null;
      try {
        return callback.apply(metalOwnedTextureFrame(frameSegment, scope));
      } catch (Throwable throwable) {
        callbackFailure = throwable;
        throw throwable;
      } finally {
        try {
          releaseMetalFrame(frameSegment, callbackFailure);
        } finally {
          scope.close();
        }
      }
    }
  }

  public void withVulkanOwnedTextureFrame(Consumer<VulkanOwnedTextureFrame> callback) {
    Objects.requireNonNull(callback, "callback");
    withVulkanOwnedTextureFrame(
        frame -> {
          callback.accept(frame);
          return null;
        });
  }

  public <T> T withVulkanOwnedTextureFrame(Function<VulkanOwnedTextureFrame, T> callback) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(callback, "callback");
    try (var arena = Arena.ofConfined()) {
      var frameSegment = RenderStructs.vulkanOwnedTextureFrame(arena);
      Status.check(
          MapLibreNativeC.mln_vulkan_owned_texture_acquire_frame(
              state.requireLive(), frameSegment));
      var scope = new FrameScope();
      Throwable callbackFailure = null;
      try {
        return callback.apply(vulkanOwnedTextureFrame(frameSegment, scope));
      } catch (Throwable throwable) {
        callbackFailure = throwable;
        throw throwable;
      } finally {
        try {
          releaseVulkanFrame(frameSegment, callbackFailure);
        } finally {
          scope.close();
        }
      }
    }
  }

  private List<QueriedFeature> queryRenderedFeaturesInternal(
      RenderedQueryGeometry geometry, RenderedFeatureQueryOptions options, boolean hasOptions) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(geometry, "geometry");
    try (var arena = Arena.ofConfined()) {
      var outResult = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_render_session_query_rendered_features(
              state.requireLive(),
              QueryStructs.renderedQueryGeometry(geometry, arena),
              hasOptions
                  ? QueryStructs.renderedFeatureQueryOptions(options, arena)
                  : MemorySegment.NULL,
              outResult));
      return QueryStructs.featureQueryResult(outResult.get(ValueLayout.ADDRESS, 0));
    }
  }

  private List<QueriedFeature> querySourceFeaturesInternal(
      String sourceId, SourceFeatureQueryOptions options, boolean hasOptions) {
    NativeAccess.ensureLoaded();
    try (var arena = Arena.ofConfined()) {
      var outResult = MemoryUtil.allocatePointer(arena);
      Status.check(
          MapLibreNativeC.mln_render_session_query_source_features(
              state.requireLive(),
              CoreStructs.stringView(Objects.requireNonNull(sourceId, "sourceId"), arena),
              hasOptions
                  ? QueryStructs.sourceFeatureQueryOptions(options, arena)
                  : MemorySegment.NULL,
              outResult));
      return QueryStructs.featureQueryResult(outResult.get(ValueLayout.ADDRESS, 0));
    }
  }

  @Override
  public void close() {
    NativeAccess.ensureLoaded();
    state.closeOnce(MapLibreNativeC::mln_render_session_destroy);
  }

  public boolean isClosed() {
    return state.isReleased();
  }

  public MapHandle map() {
    return map;
  }

  MemorySegment nativeHandle() {
    return state.requireLive();
  }

  long nativeAddress() {
    return state.address();
  }

  private static MetalOwnedTextureFrame metalOwnedTextureFrame(
      MemorySegment segment, FrameScope scope) {
    return new MetalOwnedTextureFrame(
        scope,
        mln_metal_owned_texture_frame.generation(segment),
        mln_metal_owned_texture_frame.width(segment),
        mln_metal_owned_texture_frame.height(segment),
        mln_metal_owned_texture_frame.scale_factor(segment),
        mln_metal_owned_texture_frame.frame_id(segment),
        pointer(mln_metal_owned_texture_frame.texture(segment), scope),
        pointer(mln_metal_owned_texture_frame.device(segment), scope),
        mln_metal_owned_texture_frame.pixel_format(segment));
  }

  private static VulkanOwnedTextureFrame vulkanOwnedTextureFrame(
      MemorySegment segment, FrameScope scope) {
    return new VulkanOwnedTextureFrame(
        scope,
        mln_vulkan_owned_texture_frame.generation(segment),
        mln_vulkan_owned_texture_frame.width(segment),
        mln_vulkan_owned_texture_frame.height(segment),
        mln_vulkan_owned_texture_frame.scale_factor(segment),
        mln_vulkan_owned_texture_frame.frame_id(segment),
        pointer(mln_vulkan_owned_texture_frame.image(segment), scope),
        pointer(mln_vulkan_owned_texture_frame.image_view(segment), scope),
        pointer(mln_vulkan_owned_texture_frame.device(segment), scope),
        mln_vulkan_owned_texture_frame.format(segment),
        mln_vulkan_owned_texture_frame.layout(segment));
  }

  private static NativePointer pointer(MemorySegment segment, FrameScope scope) {
    return MemoryUtil.isNull(segment)
        ? NativePointer.NULL
        : NativePointer.scoped(segment.address(), scope);
  }

  private void releaseMetalFrame(MemorySegment frameSegment, Throwable callbackFailure) {
    try {
      Status.check(
          MapLibreNativeC.mln_metal_owned_texture_release_frame(state.requireLive(), frameSegment));
    } catch (Throwable releaseFailure) {
      if (callbackFailure != null) {
        callbackFailure.addSuppressed(releaseFailure);
      } else {
        throw releaseFailure;
      }
    }
  }

  private void releaseVulkanFrame(MemorySegment frameSegment, Throwable callbackFailure) {
    try {
      Status.check(
          MapLibreNativeC.mln_vulkan_owned_texture_release_frame(
              state.requireLive(), frameSegment));
    } catch (Throwable releaseFailure) {
      if (callbackFailure != null) {
        callbackFailure.addSuppressed(releaseFailure);
      } else {
        throw releaseFailure;
      }
    }
  }
}

File: bindings/kotlin/src/jvmMain/kotlin/org/maplibre/nativeffi/render/RenderSessionHandle.kt
package org.maplibre.nativeffi.render

import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.Deferred
import org.maplibre.nativeffi.internal.lifecycle.HandleLeakCleaner
import org.maplibre.nativeffi.internal.lifecycle.HandleStateCore
import org.maplibre.nativeffi.internal.lifecycle.NativeRenderSession
import org.maplibre.nativeffi.internal.loader.NativeAccess
import org.maplibre.nativeffi.internal.status.Status
import org.maplibre.nativeffi.map.MapHandle
import org.maplibre.nativeffi.query.FeatureStateSelector
import org.maplibre.nativeffi.query.QueriedFeature
import org.maplibre.nativeffi.query.RenderedFeatureQueryOptions
import org.maplibre.nativeffi.query.RenderedQueryGeometry
import org.maplibre.nativeffi.query.SourceFeatureQueryOptions

/** Owned JVM FFM render-session control handle. */
public actual class RenderSessionHandle
internal constructor(private val ownerMap: MapHandle, private val handle: NativeRenderSession) :
  AutoCloseable {
  private val core = HandleStateCore("RenderSessionHandle", handle.raw, ownerMap)
  private val runtime = ownerMap.runtime()
  private val acquiredFrameScopes = ConcurrentHashMap.newKeySet<FrameScope>()

  init {
    HandleLeakCleaner.register(this, core.leakReport)
  }

  public actual val isClosed: Boolean
    get() = core.isReleased()

  public actual fun map(): MapHandle = ownerMap

  public actual fun capabilities(): RenderSessionCapabilities {
    NativeAccess.ensureLoaded()
    return NativeAccess.renderSessionCapabilities(requireLiveHandle())
  }

  public actual fun snapshot(): RenderSessionSnapshot {
    NativeAccess.ensureLoaded()
    return NativeAccess.renderSessionSnapshot(requireLiveHandle())
  }

  public actual fun requestFrame(demand: FrameDemand) {
    NativeAccess.ensureLoaded()
    NativeAccess.requestRenderFrame(requireLiveHandle(), demand)
  }

  public actual fun drainFrameResults(): List<RenderFrameResult> {
    NativeAccess.ensureLoaded()
    return NativeAccess.drainRenderFrameResults(requireLiveHandle())
  }

  public actual fun serviceDriverWork(maxWork: Int): Int {
    NativeAccess.ensureLoaded()
    Status.requireArgument(maxWork >= 0) { "maxWork must be non-negative" }
    return NativeAccess.serviceRenderDriverWork(requireLiveHandle(), maxWork)
  }

  public actual fun acquireFrame(): AcquiredFrameHandle? {
    NativeAccess.ensureLoaded()
    val nativeFrame = NativeAccess.acquireRenderFrame(requireLiveHandle()) ?: return null
    val scope = FrameScope()
    acquiredFrameScopes.add(scope)
    return AcquiredFrameHandle(this, nativeFrame, scope)
  }

  public actual fun resize(extent: RenderTargetExtent): Deferred<Unit> =
    NativeAccess.resizeRenderSession(requireLiveHandle(), extent)

  public actual fun setMetalSurfaceTarget(descriptor: MetalSurfaceDescriptor): Deferred<Unit> =
    NativeAccess.setMetalSurfaceTarget(requireLiveHandle(), descriptor)

  public actual fun setVulkanSurfaceTarget(descriptor: VulkanSurfaceDescriptor): Deferred<Unit> =
    NativeAccess.setVulkanSurfaceTarget(requireLiveHandle(), descriptor)

  public actual fun setOpenGLSurfaceTarget(descriptor: OpenGLSurfaceDescriptor): Deferred<Unit> =
    NativeAccess.setOpenGLSurfaceTarget(requireLiveHandle(), descriptor)

  public actual fun setMetalBorrowedTextureTarget(
    descriptor: MetalBorrowedTextureDescriptor
  ): Deferred<Unit> = NativeAccess.setMetalBorrowedTextureTarget(requireLiveHandle(), descriptor)

  public actual fun setVulkanBorrowedTextureTarget(
    descriptor: VulkanBorrowedTextureDescriptor
  ): Deferred<Unit> = NativeAccess.setVulkanBorrowedTextureTarget(requireLiveHandle(), descriptor)

  public actual fun setOpenGLBorrowedTextureTarget(
    descriptor: OpenGLBorrowedTextureDescriptor
  ): Deferred<Unit> = NativeAccess.setOpenGLBorrowedTextureTarget(requireLiveHandle(), descriptor)

  public actual fun reduceMemoryUse(): Deferred<Unit> =
    NativeAccess.renderControl(requireLiveHandle(), "mln_render_session_reduce_memory_use")

  public actual fun clearData(): Deferred<Unit> =
    NativeAccess.renderControl(requireLiveHandle(), "mln_render_session_clear_data")

  public actual fun dumpDebugLogs(): Deferred<Unit> =
    NativeAccess.renderControl(requireLiveHandle(), "mln_render_session_dump_debug_logs")

  public actual fun barrier(): Deferred<Unit> = NativeAccess.renderBarrier(requireLiveHandle())

  public actual fun detach(): Deferred<Unit> =
    NativeAccess.renderControl(requireLiveHandle(), "mln_render_session_detach")

  public actual fun setFeatureState(
    selector: FeatureStateSelector,
    value: ByteArray,
  ): Deferred<Unit> = NativeAccess.setFeatureState(requireLiveHandle(), selector, value)

  public actual fun getFeatureState(selector: FeatureStateSelector): Deferred<ByteArray> =
    NativeAccess.getFeatureState(requireLiveHandle(), selector)

  public actual fun removeFeatureState(selector: FeatureStateSelector): Deferred<Unit> =
    NativeAccess.removeFeatureState(requireLiveHandle(), selector)

  public actual fun queryRenderedFeatures(
    geometry: RenderedQueryGeometry,
    options: RenderedFeatureQueryOptions?,
  ): Deferred<List<QueriedFeature>> =
    NativeAccess.queryRenderedFeatures(requireLiveHandle(), geometry, options)

  public actual fun querySourceFeatures(
    sourceId: String,
    options: SourceFeatureQueryOptions?,
  ): Deferred<List<QueriedFeature>> =
    NativeAccess.querySourceFeatures(requireLiveHandle(), sourceId, options)

  public actual fun queryFeatureExtension(
    sourceId: String,
    feature: ByteArray,
    extension: String,
    extensionField: String,
    arguments: ByteArray?,
  ): Deferred<ByteArray> =
    NativeAccess.queryFeatureExtension(
      requireLiveHandle(),
      sourceId,
      feature,
      extension,
      extensionField,
      arguments,
    )

  public actual fun readPremultipliedRgba8(): Deferred<TextureReadback> =
    NativeAccess.textureReadback(requireLiveHandle())

  public actual fun abandon(): RenderAbandonResult {
    NativeAccess.ensureLoaded()
    val result = NativeAccess.abandonRenderSession(requireLiveHandle())
    acquiredFrameScopes.forEach(FrameScope::close)
    return result
  }

  public actual override fun close() {
    NativeAccess.ensureLoaded()
    core.closeOnce(destroy = { NativeAccess.destroyRenderSession(handle) })
  }

  internal fun frameReleased(scope: FrameScope) {
    acquiredFrameScopes.remove(scope)
  }

  private fun requireLiveHandle(): NativeRenderSession {
    core.requireLive()
    return handle
  }

  internal companion object {
    fun attachMetalOwnedTexture(
      map: MapHandle,
      descriptor: MetalOwnedTextureDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(map, NativeAccess.attachMetalOwnedTexture(map.nativeHandle(), descriptor, options))

    fun attachMetalBorrowedTexture(
      map: MapHandle,
      descriptor: MetalBorrowedTextureDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(
        map,
        NativeAccess.attachMetalBorrowedTexture(map.nativeHandle(), descriptor, options),
      )

    fun attachVulkanOwnedTexture(
      map: MapHandle,
      descriptor: VulkanOwnedTextureDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(
        map,
        NativeAccess.attachVulkanOwnedTexture(map.nativeHandle(), descriptor, options),
      )

    fun attachVulkanBorrowedTexture(
      map: MapHandle,
      descriptor: VulkanBorrowedTextureDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(
        map,
        NativeAccess.attachVulkanBorrowedTexture(map.nativeHandle(), descriptor, options),
      )

    fun attachOpenGLOwnedTexture(
      map: MapHandle,
      descriptor: OpenGLOwnedTextureDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(
        map,
        NativeAccess.attachOpenGLOwnedTexture(map.nativeHandle(), descriptor, options),
      )

    fun attachOpenGLBorrowedTexture(
      map: MapHandle,
      descriptor: OpenGLBorrowedTextureDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(
        map,
        NativeAccess.attachOpenGLBorrowedTexture(map.nativeHandle(), descriptor, options),
      )

    fun attachMetalSurface(
      map: MapHandle,
      descriptor: MetalSurfaceDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(map, NativeAccess.attachMetalSurface(map.nativeHandle(), descriptor, options))

    fun attachVulkanSurface(
      map: MapHandle,
      descriptor: VulkanSurfaceDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(map, NativeAccess.attachVulkanSurface(map.nativeHandle(), descriptor, options))

    fun attachOpenGLSurface(
      map: MapHandle,
      descriptor: OpenGLSurfaceDescriptor,
      options: RenderSessionAttachOptions,
    ): RenderSessionAttachment =
      attachment(map, NativeAccess.attachOpenGLSurface(map.nativeHandle(), descriptor, options))

    private fun attachment(
      map: MapHandle,
      native: Pair<NativeRenderSession, Deferred<Unit>>,
    ): RenderSessionAttachment {
      NativeAccess.ensureLoaded()
      val session = RenderSessionHandle(map, native.first)
      return try {
        RenderSessionAttachment(session, retainSessionUntilComplete(session, native.second))
      } catch (error: Throwable) {
        runCatching { session.abandon() }
        runCatching { session.close() }
        throw error
      }
    }
  }
}

/** JVM FFM acquired-frame lease. */
public actual class AcquiredFrameHandle
internal constructor(
  private val session: RenderSessionHandle,
  frame: Long,
  private val scope: FrameScope,
) {
  @Volatile private var nativeFrame: Long = frame

  public actual val isReleased: Boolean
    get() = nativeFrame == 0L

  public actual fun result(): RenderFrameResult =
    NativeAccess.acquiredFrameResult(requireAccessibleFrame())

  public actual fun producerSync(): GpuSync =
    NativeAccess.acquiredFrameProducerSync(requireAccessibleFrame())

  public actual fun metalTexture(): MetalOwnedTextureFrame =
    NativeAccess.acquiredMetalTexture(requireAccessibleFrame(), scope)

  public actual fun vulkanTexture(): VulkanOwnedTextureFrame =
    NativeAccess.acquiredVulkanTexture(requireAccessibleFrame(), scope)

  public actual fun openGLTexture(): OpenGLOwnedTextureFrame =
    NativeAccess.acquiredOpenGLTexture(requireAccessibleFrame(), scope)

  @Synchronized
  public actual fun release(consumerCompletion: GpuSync) {
    val frame = requireFrame()
    val remainingFrame = NativeAccess.releaseAcquiredFrame(frame, consumerCompletion)
    check(remainingFrame == 0L) { "native frame release did not consume the handle" }
    nativeFrame = 0L
    scope.close()
    session.frameReleased(scope)
  }

  private fun requireAccessibleFrame(): Long {
    scope.ensureActive()
    return requireFrame()
  }

  private fun requireFrame(): Long {
    val frame = nativeFrame
    if (frame == 0L) throw Status.invalidState("AcquiredFrameHandle is released")
    return frame
  }
}
